test/decl/var/usage.swift

```diff
@@ -563,3 +563,11 @@ func testUselessCastWithInvalidParam(foo: Any?) -> Int {
   if let bar = foo as? Foo { return 42 } // expected-warning {{value 'bar' was defined but never used; consider replacing with boolean test}} {{6-16=}} {{20-23=is}}
   else { return 54 }
 }
+
+// https://github.com/swiftlang/swift/issues/72811
+func testEnumeratedForLoop(a: [Int]) {
+  for var (b, c) in a.enumerated() {  // expected-warning {{variable 'b' was never mutated; consider changing to 'case let' constant}}
```

Comment:
I think we either need to a separate diagnostic here, if fix-it cannot be produced, which would suggest exactly how to do this because the correct form is `case (let b, var c)` which not not be obvious from the diagnostic. I'm also not sure why `FixItLoc` would be invalid here, there might be a bug in `getFixItLocForVarToLet` because `var` applies to both elements and it should be possible to produce a fix-it in this case.

Reply:
hi, thanks for the input. indeed, I find it strange that `FixItLoc` is invalid, let me take closer look at `getFixItLocForVarToLet`. Otherwise, I would create separate diagnostic with the correct syntax.

Reply:
Hi @xedin ,

I have updated diagnostic message.

For `getFixItLocForVarToLet`, it is not possible because it is using value binding pattern to detect the usage of `var` and tuple pattern does not contain such pattern in for-in.

Alternatively, I noticed there is prior check `isVarDeclPartOfPBDThatHadSomeMutation` that specifically do not generate such mutated warning for tuple use case. Currently, it does not handle tuple pattern in for-in. I do think this is cleaner approach and more aligned with existing behavior for tuple partial mutation, e.g. following code has no such warning for `y`
```
var (x, y) = (1, 2)
x = 3
let _ = x
let _ = y
```
I stop short of making the changes though, as it differ from the solution discussion in #72811 and in this PR.

Reply:
I like the new message better but it still needs some work because `it` in there refers to the variable but suggestion to to change the whole pattern, I think it needs to be rephrased to indicate that the whole pattern is going to change to do this.

Reply:
Hi @xedin , good point. I have pushed the improved message. Let me know what do you think.